// react-motion-menu/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _item = require('./item');

var _item2 = _interopRequireDefault(_item);

var _button = require('./button');

var _button2 = _interopRequireDefault(_button);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var MotionMenu = function (_Component) {
  _inherits(MotionMenu, _Component);

  function MotionMenu(props) {
    _classCallCheck(this, MotionMenu);

    var _this = _possibleConstructorReturn(this, (MotionMenu.__proto__ || Object.getPrototypeOf(MotionMenu)).call(this, props));

    _this.state = {
      itemNumber: 1,
      status: 'idle'
    };
    _this.items = [];
    _this.onOpenEnd = _this.onOpenEnd.bind(_this);
    _this.onCloseEnd = _this.onCloseEnd.bind(_this);
    _this.onClick = _this.onClick.bind(_this);
    return _this;
  }

  _createClass(MotionMenu, [{
    key: 'onOpenEnd',
    value: function onOpenEnd(name) {
      if (this.state.action !== 'open') return;
      if (this.state.itemNumber < this.props.children.length) {
        this.items[this.state.itemNumber].start();
        this.setState({
          itemNumber: this.state.itemNumber + 1
        });
        return;
      }
      if (name === 'item' + (this.props.children.length - 1)) {
        this.props.onOpen();
      }
    }
  }, {
    key: 'onCloseEnd',
    value: function onCloseEnd(name) {
      if (this.state.action === 'open') return;
      if (name === 'item1') {
        this.props.onClose();
      }
      if (this.state.itemNumber > 1) {
        if (name === 'item1') {
          this.props.onClose();
        }
        this.setState({
          itemNumber: this.state.itemNumber - 1
        });
      }
    }
  }, {
    key: 'onClick',
    value: function onClick() {
      if (this.state.action === 'open') {
        this.closeItems();
      } else {
        this.openItem();
      }
    }
  }, {
    key: 'getDistance',
    value: function getDistance(i) {
      return this.props.wing ? (parseInt(i / 2, 10) + 1) * this.props.margin * (i % 2 || -1) : this.props.margin * (i + 1);
    }
  }, {
    key: 'getX',
    value: function getX(i, x) {
      var _props = this.props,
          type = _props.type,
          margin = _props.margin,
          children = _props.children;

      if (type === 'horizontal') {
        return this.getDistance(i) + x;
      }
      if (type === 'circle') {
        return x + margin * Math.cos(Math.PI * 2 * i / (children.length - 1));
      }
      return x;
    }
  }, {
    key: 'getY',
    value: function getY(i, y) {
      var _props2 = this.props,
          type = _props2.type,
          margin = _props2.margin,
          children = _props2.children;

      if (type === 'vertical') {
        return this.getDistance(i) + y;
      }
      if (type === 'circle') {
        return y + margin * Math.sin(Math.PI * 2 * i / (children.length - 1));
      }
      return y;
    }
  }, {
    key: 'getItems',
    value: function getItems() {
      var _this2 = this;

      var _props3 = this.props,
          x = _props3.x,
          y = _props3.y,
          bumpy = _props3.bumpy;

      return Array.from(Array(this.state.itemNumber).keys()).reverse().map(function (i) {
        return _react2.default.createElement(
          _item2.default,
          {
            key: i,
            ref: function ref(c) {
              _this2.items[i + 1] = c;
            },
            name: 'item' + (i + 1),
            onOpenAnimationEnd: _this2.onOpenEnd,
            onCloseAnimationEnd: _this2.onCloseEnd,
            x: _this2.getX(i, x),
            y: _this2.getY(i, y),
            bumpy: bumpy,
            openSpeed: _this2.props.openSpeed,
            reverse: _this2.props.reverse,
            type: _this2.props.type
          },
          _this2.props.children[i + 1]
        );
      });
    }
  }, {
    key: 'closeItems',
    value: function closeItems() {
      var _this3 = this;

      this.setState({ action: 'close' });
      this.button.reverse();
      Array.from(Array(this.state.itemNumber).keys()).reverse().forEach(function (i) {
        return _this3.items[i + 1].reverse();
      });
    }
  }, {
    key: 'close',
    value: function close() {
      if (this.state.action !== 'open') return;
      this.closeItems();
    }
  }, {
    key: 'open',
    value: function open() {
      if (this.state.action === 'open') return;
      this.openItem();
    }
  }, {
    key: 'openItem',
    value: function openItem() {
      this.setState({ action: 'open' });
      this.button.start();
      this.items[this.state.itemNumber].start();
    }
  }, {
    key: 'render',
    value: function render() {
      return _react2.default.createElement(
        'div',
        {
          style: this.props.style,
          className: this.props.className
        },
        _react2.default.createElement(
          'div',
          { style: { position: 'relative' } },
          this.menuButton,
          this.getItems()
        )
      );
    }
  }, {
    key: 'menuButton',
    get: function get() {
      var _this4 = this;

      return _react2.default.createElement(
        _button2.default,
        {
          ref: function ref(c) {
            _this4.button = c;
          },
          onClick: this.onClick,
          x: this.props.x,
          y: this.props.y,
          bumpy: this.props.bumpy
        },
        this.props.children[0]
      );
    }
  }]);

  return MotionMenu;
}(_react.Component);

MotionMenu.propTypes = {
  margin: _react.PropTypes.number.isRequired,
  type: _react.PropTypes.oneOf(['horizontal', 'vertical', 'circle']).isRequired,
  wing: _react.PropTypes.bool,
  x: _react.PropTypes.number,
  y: _react.PropTypes.number,
  onClose: _react.PropTypes.func,
  onOpen: _react.PropTypes.func,
  className: _react.PropTypes.string,
  bumpy: _react.PropTypes.bool,
  openSpeed: _react.PropTypes.number,
  reverse: _react.PropTypes.bool
};
MotionMenu.defaultProps = {
  x: 0,
  y: 0,
  style: {},
  onClose: function onClose() {},
  onOpen: function onOpen() {},
  bumpy: true,
  openSpeed: 60,
  reverse: false
};
exports.default = MotionMenu;
