// react-motion-menu/lib/item.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactMotion = require('react-motion');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var createSmoothParams = function createSmoothParams(_ref) {
  var x = _ref.x,
      y = _ref.y;
  return [{
    scaleX: (0, _reactMotion.spring)(0, { stiffness: 1500, damping: 100 }),
    scaleY: (0, _reactMotion.spring)(0, { stiffness: 1500, damping: 100 }),
    x: (0, _reactMotion.spring)(x, { stiffness: 1500, damping: 50 }),
    y: (0, _reactMotion.spring)(y, { stiffness: 1500, damping: 50 })
  }, {
    scaleX: (0, _reactMotion.spring)(0.5, { stiffness: 120, damping: 20 }),
    scaleY: (0, _reactMotion.spring)(0.5, { stiffness: 120, damping: 20 }),
    x: (0, _reactMotion.spring)(x, { stiffness: 120, damping: 20 }),
    y: (0, _reactMotion.spring)(y, { stiffness: 120, damping: 20 })
  }, {
    scaleX: (0, _reactMotion.spring)(1, { stiffness: 120, damping: 20 }),
    scaleY: (0, _reactMotion.spring)(1, { stiffness: 120, damping: 20 }),
    x: (0, _reactMotion.spring)(x, { stiffness: 120, damping: 20 }),
    y: (0, _reactMotion.spring)(y, { stiffness: 120, damping: 20 })
  }];
};

var createBumpyParams = function createBumpyParams(x, y) {
  return [{
    scaleX: (0, _reactMotion.spring)(0, { stiffness: 1500, damping: 100 }),
    scaleY: (0, _reactMotion.spring)(0, { stiffness: 1500, damping: 100 }),
    x: (0, _reactMotion.spring)(x, { stiffness: 1500, damping: 50 }),
    y: (0, _reactMotion.spring)(y, { stiffness: 1500, damping: 50 })
  }, {
    scaleX: (0, _reactMotion.spring)(1.6, { stiffness: 1500, damping: 150 }),
    scaleY: (0, _reactMotion.spring)(0.7, { stiffness: 1500, damping: 150 }),
    x: (0, _reactMotion.spring)(x, { stiffness: 1500, damping: 100 }),
    y: (0, _reactMotion.spring)(y, { stiffness: 1500, damping: 100 })
  }, {
    scaleX: (0, _reactMotion.spring)(1, { stiffness: 1500, damping: 18 }),
    scaleY: (0, _reactMotion.spring)(1, { stiffness: 1500, damping: 18 }),
    x: (0, _reactMotion.spring)(x, { stiffness: 1500, damping: 100 }),
    y: (0, _reactMotion.spring)(y, { stiffness: 1500, damping: 100 })
  }];
};

var MenuItem = function (_Component) {
  _inherits(MenuItem, _Component);

  function MenuItem(props) {
    _classCallCheck(this, MenuItem);

    var _this = _possibleConstructorReturn(this, (MenuItem.__proto__ || Object.getPrototypeOf(MenuItem)).call(this, props));

    _this.timerIds = [];
    _this.state = {
      sequence: 0
    };

    _this.sequenceParams = _this.props.bumpy ? createBumpyParams(props) : createSmoothParams(props);
    return _this;
  }

  _createClass(MenuItem, [{
    key: 'start',
    value: function start() {
      var _this2 = this;

      this.timerIds[1] = setTimeout(function () {
        _this2.setState({ sequence: 1 });
        _this2.timerIds[1] = null;
      }, this.props.openSpeed);

      this.timerIds[2] = setTimeout(function () {
        _this2.setState({ sequence: 2 });
        _this2.timerIds[2] = null;
        _this2.props.onOpenAnimationEnd(_this2.props.name);
      }, this.props.openSpeed);
    }
  }, {
    key: 'reverse',
    value: function reverse() {
      var _this3 = this;

      this.timerIds.forEach(function (id) {
        if (id) clearTimeout(id);
      });
      this.timerIds[0] = setTimeout(function () {
        _this3.timerIds[0] = null;
        _this3.props.onCloseAnimationEnd(_this3.props.name);
      }, 100);
      this.setState({ sequence: 0 });
    }
  }, {
    key: 'render',
    value: function render() {
      var _this4 = this;

      var _props = this.props,
          x = _props.x,
          y = _props.y,
          reverse = _props.reverse,
          type = _props.type;

      var newX = void 0;
      var newY = void 0;
      if (reverse) {
        newX = -1 * x;
        newY = type === 'vertical' ? -1 * y : y;
      } else {
        newX = x;
        newY = y;
      }
      if (!this.props.children) return null;
      return _react2.default.createElement(
        _reactMotion.Motion,
        { style: this.sequenceParams[this.state.sequence] },
        function (_ref2) {
          var scaleX = _ref2.scaleX,
              scaleY = _ref2.scaleY;
          return (0, _react.cloneElement)(_this4.props.children, _extends({}, _this4.props.children.props || {}, {
            style: _extends({}, _this4.props.children.props && _this4.props.children.props.style || {}, {
              transform: 'translate3d(' + newX + 'px, ' + newY + 'px, 0) scaleX(' + scaleX + ') scaleY(' + scaleY + ')',
              WebkitTransform: 'translate3d(' + newX + 'px, ' + newY + 'px, 0) scaleX(' + scaleX + ') scaleY(' + scaleY + ')',
              position: 'absolute'
            })
          }));
        }
      );
    }
  }]);

  return MenuItem;
}(_react.Component);

MenuItem.propTypes = {
  x: _react.PropTypes.number.isRequired,
  y: _react.PropTypes.number.isRequired,
  name: _react.PropTypes.string.isRequired,
  onOpenAnimationEnd: _react.PropTypes.func,
  onCloseAnimationEnd: _react.PropTypes.func,
  bumpy: _react.PropTypes.bool.isRequired,
  openSpeed: _react.PropTypes.number.isRequired,
  reverse: _react.PropTypes.bool.isRequired,
  type: _react.PropTypes.oneOf(['horizontal', 'vertical', 'circle']).isRequired
};
MenuItem.defaultProps = {
  onOpenAnimationEnd: function onOpenAnimationEnd() {},
  onCloseAnimationEnd: function onCloseAnimationEnd() {}
};
exports.default = MenuItem;
