// resonance/Animate/Animate.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _transition = require('../core/transition');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

//  weak
/* eslint max-len: "off" */

var Animate = function (_Component) {
  (0, _inherits3.default)(Animate, _Component);

  function Animate() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, Animate);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = Animate.__proto__ || (0, _getPrototypeOf2.default)(Animate)).call.apply(_ref, [this].concat(args))), _this), _this.state = _this.props.start(_this.props.data), _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(Animate, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var _props = this.props,
          data = _props.data,
          enter = _props.enter;


      if (enter && typeof enter === 'function') {
        _transition.transition.call(this, enter(data));
      }
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(next) {
      if (next.data !== this.props.data) {
        this.update(next);
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      _transition.stop.call(this);
    }
  }, {
    key: 'update',
    value: function update(props) {
      var data = props.data,
          update = props.update;

      _transition.transition.call(this, update(data));
    }
  }, {
    key: 'render',
    value: function render() {
      var renderedChildren = this.props.children(this.props.data, this.state);
      return renderedChildren && _react2.default.Children.only(renderedChildren);
    }
  }]);
  return Animate;
}(_react.Component);

Animate.defaultProps = {
  enter: function enter() {},
  update: function update() {}
};
exports.default = Animate;
